src/THREE.MeshLine.js

```diff
@@ -1,782 +1,714 @@
-	this.far = source.far;
-	this.dashArray.copy( source.dashArray );
-	this.dashOffset.copy( source.dashOffset );
-	this.dashRatio.copy( source.dashRatio );
-	this.useDash = source.useDash;
-	this.visibility = source.visibility;
-	this.alphaTest = source.alphaTest;
-	this.repeat.copy( source.repeat );
-
-	return this;
-
-};
-
-if( typeof exports !== 'undefined' ) {
-	if( typeof module !== 'undefined' && module.exports ) {
-		exports = module.exports = { MeshLine: MeshLine, MeshLineMaterial: MeshLineMaterial };
-	}
-	exports.MeshLine = MeshLine;
-	exports.MeshLineMaterial = MeshLineMaterial;
-}
-else {
-	root.MeshLine = MeshLine;
-	root.MeshLineMaterial = MeshLineMaterial;
-}
-
-}).call(this);
+  'use strict'
+
+  var root = this
+
+  var has_require = typeof require !== 'undefined'
+
+  var THREE = root.THREE || (has_require && require('three'))
+  if (!THREE) throw new Error('MeshLine requires three.js')
+
+  function MeshLine() {
+    THREE.BufferGeometry.call(this)
+    this.type = 'MeshLine'
+
+    this.positions = []
+
+    this.previous = []
+    this.next = []
+    this.side = []
+    this.width = []
+    this.indices_array = []
+    this.uvs = []
+    this.counters = []
+    this._vertices = []
+    this._bufferArray = []
+
+    this.widthCallback = null
+
+    // Used to raycast
+    this.matrixWorld = new THREE.Matrix4()
+
+    // to support previous api
+    Object.defineProperties(this, {
+      geometry: {
+        enumerable: true,
+        get: function() {
+          return this
+        },
+        set: function(value) {
+          this.setFromGeometry(value)
+        },
+      },
+      vertices: {
+        enumerable: true,
+        get: function() {
+          return this._vertices
+        },
+        set: function(value) {
+          this.setVertices(value)
+        },
+      },
+      bufferArray: {
+        enumerable: true,
+        get: function() {
+          return this._bufferArray
+        },
+        set: function(value) {
+          this.setBufferArray(value)
+        },
+      },
+    })
+  }
+
+  MeshLine.prototype = Object.create(THREE.BufferGeometry.prototype)
+  MeshLine.prototype.constructor = MeshLine
+  MeshLine.prototype.isMeshLine = true
+
+  MeshLine.prototype.setMatrixWorld = function(matrixWorld) {
+    this.matrixWorld = matrixWorld
+  }
+
+  MeshLine.prototype.setFromGeometry = function(g, c) {
+    if (g instanceof THREE.Geometry) {
+      this.setVertices(g.vertices, c)
+    }
+    if (g instanceof THREE.BufferGeometry) {
+      this.setBufferArray(g.getAttribute('position').array, c)
+    }
+    if (g instanceof Float32Array || g instanceof Array) {
+      // to support previous api
+      this.setBufferArray(g, c)
+    }
+  }
+
+  // to support previous api
+  MeshLine.prototype.setGeometry = function(g, c) {
+    this.setFromGeometry(g, c)
+  }
+
+  MeshLine.prototype.setVertices = function(vts, wcb) {
+    this._vertices = vts
+    this.widthCallback = wcb || this.widthCallback
```

Comment:
`this.widthCallback` is capable of being set prior to `setBufferArray` being called. If `wcb` is undefined it will override the existing value. Adding the or operator will keep the existing value if defined prior.

Reply:
Declarative architectures will likely set `widthCallback` via `this.widthCallback = func`, rather than using `setVertices` or `setBufferArray`

Reply:
I'm confused about this part. 

Initially, there was only `setGeometry()` to process the lines. But now we also have `setVertices` & `setBufferArray` with some duplicated piece of code and who almost do the same.

I really like the idea to have a more performant solution like you offer into `setVertices`.
But to have something more consistent, I would prefer **only one new method controlling the entire process**. I would also **avoid using the word `vertices`** who can be misunderstanding. Indeed, the array given will not be the real list vertices.

Please take a look of my new suggestion who try to solve this part with a larger update.

Thanks!

Reply:
I've accidentally pushed some commits into your PR with my update :O 
I don't know why I was allowed to do it, sorry about that I wanted first to do a PR too.

Let me know your thought about it!

Reply:
All good, who knows how this all works 🤷🏼‍♂️ 

I got a little mixed up merging your commits, but it's essentially the same as you had it.
Just added back in `this.widthCallback = wcb || this.widthCallback` because if it's set prior setPoints can override the value with undefined.

Reply:
Amazing!

But I'm still not sure about the `this.widthCallback = wcb || this.widthCallback`.

Please take a look of the lines `l.220 - l.221`: 
```
if (this.widthCallback) w = this.widthCallback(j / (l - 1))
else w = 1
```

It means if the callback is doesn't exist (`undefined`),  the `width` will be set to 1 by default. This default behaviour might be applied each time you don't want to use (or reuse) `setPoint` with a callback.

If we add your change, the second time `setPoint` will be used (this time voluntary without callback) , the previous callback will be called anyway. Even if you don't want it or don't plane to reuse it.

Finally, `setPoint` is not really made to be called a second time. It should be really rare to have to update the vertices of an existing Mesh.

It's why I'll prefer to avoid this update. I don't see any advantage of it for now. Please let me know your thought.

Reply:
With declarative architectures, you will basically be doing the equivalent of

```js
meshline.widthCallback = widthFunc
meshline.points = points // here meshline.widthCallback will now be set to undefined when setPoints is called
```

I think I understand where you're coming from now. That with my code, there's no way to set `widthCallback` to null using the traditional `setPoints` function. What if we update the `points` & `geometry` setters and getters for declarative architectures instead? Since you're specifically editing points or geometry, you shouldn't expect `widthCallback` to change.

```js
points: {
        enumerable: true,
        get: function() {
          return this._points
        },
        set: function(value) {
          this.setPoints(value, this.widthCallback)
        },
      },
```

Reply:
I've also just realised there's nothing to handle `setPoints` being directly given an array of vector3's

Reply:
Ok, I've updated `setPoints` to handle being given an array of `Vector3`s
As well as an alternative solution to `this.widthCallback = wcb || this.widthCallback`

Reply:
For `setPoint`, I don't think we need to have multiple options.

There is not too much difference about building an array of `THREE.Vector3` or an array of numbers.
Also, build an array of numbers looks more optimal than build each Vector3 Object.

But I'm ok to let it like that! I like the way you updated the code of  `setGeometry` & `setPoints`.